### ISSUE TITLE
1.5.21 fails self-tests on s390x (test_stats_prefix_dump)

### ISSUE THREAD

**Issue by cpaelzer** [memcached/memcached#606]
Hi,
[Debian](ttps://buildd.debian.org/status/fetch.php?pkg=memcached&arch=s390x&ver=1.5.21-1&stamp=1580041692&raw=0) and [Ubuntu](https://launchpadlibrarian.net/462233770/buildlog_ubuntu-focal-s390x.memcached_1.5.21-1_BUILDING.txt.gz) face the same build issue on s390x in version 1.5.21.

Tests break with:
```
...
ok 10 - stats_prefix_record_set
  testapp: testapp.c:355: test_stats_prefix_dump: Assertion `strcmp(expected, stats_prefix_dump(&length)) == 0' failed.
```

I checked - the issue occurs on git master and the plain 1.5.21 tarball - so it is not related to the distributions Delta (if any).

I don't really know what to look out for (let me know, I can keep the test container up).
What I saw with gdb was:
```
root@f-memcached:~/memcached# gdb ./testapp 
(gdb) b testapp.c:355
Breakpoint 1 at 0x42e4: file testapp.c, line 355.
(gdb) run
...
Breakpoint 1, test_stats_prefix_dump () at testapp.c:355
355         assert(strcmp(expected, stats_prefix_dump(&length)) == 0);
(gdb) p expected
$1 = 0x2aa0000b418 "PREFIX abc get 2 hit 1 set 1 del 1\r\nPREFIX def get 0 hit 0 set 0 del 1\r\nEND\r\n"
(gdb) p length
$2 = 41
(gdb) p &length
$3 = (int *) 0x3fffffff1f0
(gdb) b strcmp
Breakpoint 2 at 0x3fffdd27a10: file ../sysdeps/s390/strcmp-vx.S, line 44.
(gdb) s
stats_prefix_dump (length=0x3fffffff1f0) at stats_prefix.c:138
138         STATS_LOCK();
(gdb) p length
$4 = (int *) 0x3fffffff1f0
(gdb) p *length
$5 = 41
(gdb) n
139         size = strlen(format) + total_prefix_size +
(gdb) 
143         buf = malloc(size);
(gdb) 
144         if (NULL == buf) {
(gdb) 
152             for (pfs = prefix_stats[i]; NULL != pfs; pfs = pfs->next) {
(gdb) p buf
$6 = 0x2aa00013ab0 ""
(gdb) p size
$7 = 277
(gdb) fin
Run till exit from #0  stats_prefix_dump (length=0x3fffffff1f0) at stats_prefix.c:138
0x000002aa000042ee in test_stats_prefix_dump () at testapp.c:355
355         assert(strcmp(expected, stats_prefix_dump(&length)) == 0);
Value returned is $9 = 0x2aa00013ab0 "PREFIX def get 0 hit 0 set 0 del 1\r\nPREFIX abc get 2 hit 1 set 1 del 1\r\nEND\r\n"
```

So it compares:
- expected "PREFIX abc get 2 hit 1 set 1 del 1\r\nPREFIX def get 0 hit 0 set 0 del 1\r\nEND\r\n"
- returned "PREFIX def get 0 hit 0 set 0 del 1\r\nPREFIX abc get 2 hit 1 set 1 del 1\r\nEND\r\n"

You can see how the order is inverted - that might be a classic s390x big-endian issue after all.
Up to you - let me know if you have questions or anything to try.

**Comment by dormando** [comment]
Thanks for the report!

Those tests likely hadn't ever run before so it's not a big surprise they'd fail on some platforms. To confirm, this is new as of 1.5.21, not .20?

@kanak - feel like taking a look at this? Since you fixed them up once already :)

**Comment by kanak** [comment]
Happy to take a look. Some assertions depend on the iteration order of the hash table, which isn't great, and is likely the cause of this problem.

**Comment by kanak** [comment]
@cpaelzer : I think I have a fix for that issue here https://github.com/kanak/memcached/tree/fix_stats_prefix_test  . Do you mind applying the patch and seeing if it fixes the issue?

I don't have a big-endian machine to verify it myself :(.

**Comment by cpaelzer** [comment]
@kanak happy to try ... just give me a few minutes.

About s390x machines I'm happy to have access in our company. I think you might try [LinuxOneCloud](https://developer.ibm.com/linuxone/) - not sure but they once had free trials.

**Comment by cpaelzer** [comment]
I was trying 0ffde456 of your branch `kanak/fix_stats_prefix_test.`

With that it passed the formerly failing test ... it is still ongoing with the further test - I'll report later if all others worked as well.

**Comment by cpaelzer** [comment]
Your commit fixes the reported issue, but there are other issues later on that I hit now.

Might be a consequence of your fix (unlikely) or just formerly hidden behind the test that broke earlier.
I'll attach a (long) log in a minute ...

**Comment by cpaelzer** [comment]
The test spew out _a lot_ of 'x' chars for a lot of mismatches.
I attached a full output, maybe it means something to someone here.

(gzip as the raw txt is ~33M)

[memcached-1.5.21-s390x-test.txt.gz](https://github.com/memcached/memcached/files/4140184/memcached-1.5.21-s390x-test.txt.gz)

**Comment by dormando** [comment]
@cpaelzer sorry if I misread, but were any previous versions passing tests?
If you re-run the tests a few times, do they pass or is it always flaky? We've had #381 open for a long time which might be a similar issue, but I don't have access to such a system so I've not tried to fix it myself.

Not sure if you were trying with `--enable-extstore` or not.

also yes, the text spew on failures for some of those tests sucks :( Sorry I haven't fixed yet.

**Comment by cpaelzer** [comment]
@dormando I was just running the default:
  
    ./autogen.sh && ./configure && make && make test

The configure output isn't very verbose what the default of extstore is in that case, but in the tests later you can see

    t/extstore.t ................ skipped: extstore not enabled

Therefore we can assume `--enable-extstore` isn't playing a role here at the moment.

**Comment by cpaelzer** [comment]
I have run it a few times now, it seems reliably failing - not a flaky result.

For versions - yes everything up to 1.5.20 worked fine.

**Comment by cpaelzer** [comment]
Uh wait a minute - it seems I spoke too soon.
I was verifying it on a new system and there are indeed I have non-failing tests (with the fix by @kanak applied).

So we have:

- without kanaks fix reliably broken on the reported issue
- my old loop failed reliably
- I have taken a new system today and re-ran kanaks branch and there it seems to work
  - on this I even have got 3/3 good results now

Anyway - that makes me think I might have had some bad artifact blocking my tests last week and with the fix by @kanak I might be actually good now.

Let me throw 1.5.21 + the fix into the real Ubuntu build infrastructure ...

**Comment by cpaelzer** [comment]
Yeah, this worked to build well on [all architectures](https://launchpad.net/~ci-train-ppa-service/+archive/ubuntu/3910/+packages) now.

I'd appreciate if the change by @kanak could be merged to make it "official"

**Comment by dormando** [comment]
That's the plan. @kanak mind opening a PR? Not sure I saw one for the branch.

**Comment by cpaelzer** [comment]
FYI Debian has released 1.5.22 based code now, I rebased and it works there on all architectures as well.

**Comment by dormando** [comment]
Great! I was offline for a while. Looks like the change doesn't apply to the `next` branch  at all, so I'll give @kanak a chance to fix that up before merging. Next release should be soon.

**Comment by dormando** [comment]
Fix out in 1.6.0

**Comment by cpaelzer** [comment]
Thank you @dormando
